File: id/id/src/main/java/io/piano/android/id/PianoIdJavascriptDelegate.java
package io.piano.android.id;

import android.webkit.JavascriptInterface;

import java.lang.ref.WeakReference;

class PianoIdJavascriptDelegate implements PianoIdJsInterface {
    private final WeakReference<PianoIdJsInterface> reference;

    PianoIdJavascriptDelegate(PianoIdJsInterface delegate) {
        this.reference = new WeakReference<>(delegate);
    }

    @JavascriptInterface
    public void socialLogin(String payload) {
        PianoIdJsInterface delegate = reference.get();
        if (delegate != null)
            delegate.socialLogin(payload);
    }

    @JavascriptInterface
    public void loginSuccess(String payload) {
        PianoIdJsInterface delegate = reference.get();
        if (delegate != null)
            delegate.loginSuccess(payload);
    }

    @JavascriptInterface
    public void cancel() {
        PianoIdJsInterface delegate = reference.get();
        if (delegate != null)
            delegate.cancel();
    }
}

File: id/id/src/main/java/io/piano/android/id/PianoIdJavascriptDelegate.kt
package io.piano.android.id

import android.webkit.JavascriptInterface
import java.lang.ref.WeakReference

internal class PianoIdJavascriptDelegate(
    internalDelegate: PianoIdJsInterface,
    publicDelegate: PianoIdJs?
) : PianoIdJsInterface, PianoIdJs {
    private val internalJsReference: WeakReference<PianoIdJsInterface> = WeakReference(internalDelegate)
    private val publicJsReference: WeakReference<PianoIdJs?> = WeakReference(publicDelegate)

    @JavascriptInterface
    override fun socialLogin(payload: String?) {
        internalJsReference.get()?.socialLogin(payload)
    }

    @JavascriptInterface
    override fun registerSuccess(payload: String?) {
        internalJsReference.get()?.registerSuccess(payload)
    }

    @JavascriptInterface
    override fun loginSuccess(payload: String?) {
        internalJsReference.get()?.loginSuccess(payload)
    }

    @JavascriptInterface
    override fun cancel() {
        internalJsReference.get()?.cancel()
    }

    @JavascriptInterface
    override fun customEvent(eventData: String) {
        publicJsReference.get()?.customEvent(eventData)
    }
}
